File: examples/adapters/python/mpc_examples_adapters/controllers/mpc_trajectory_controller.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Acados trajectory-MPC adapter wrapping :class:`mpc_acados_trajectory.MPC`.

Python mirror of
``examples/adapters/controllers/mpc_trajectory/src/mpc_trajectory_controller.cpp``.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import math
import os
import time
from dataclasses import dataclass
from typing import List, Optional

import numpy as np
import yaml
from mavpy.model import State
from mpc_acados_trajectory import MPC
from mpc_acados_trajectory.utils.mpc_yaml import configure_mpc_from_yaml

from mpc_examples_framework import (
    ControlCommand,
    ExampleConfig,
    IController,
    ReferenceField,
    ReferenceSample,
    make_mask,
)


@dataclass
class MpcTrajectoryConfig:
    """Parsed configuration for :class:`MpcTrajectoryController`."""

    mpc_yaml_path: str = ''


def _yaw_to_quat(yaw: float) -> np.ndarray:
    half = 0.5 * yaw
    return np.array([math.cos(half), 0.0, 0.0, math.sin(half)], dtype=float)


def _read_ocp_json_file(path: str) -> str:
    if not os.path.isfile(path):
        raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
    with open(path, 'r') as f:
        root = yaml.safe_load(f)
    if not isinstance(root, dict):
        raise ValueError('mpc config: root must be a mapping.')
    controller = root.get('controller')
    if not isinstance(controller, dict):
        raise ValueError("mpc config: 'controller' must be a mapping.")
    ocp_path = controller.get('ocp_json_file_path')
    if not isinstance(ocp_path, str) or not ocp_path:
        raise ValueError("mpc config: 'controller.ocp_json_file_path' is required.")
    return ocp_path


class MpcTrajectoryController(IController):
    """IController adapter around the trajectory variant of the acados MPC.

    Horizon size: ``N_horizon + 1``. The generator is expected to supply
    position + velocity + acceleration at every stage; the adapter forwards
    each stage reference to the solver verbatim.
    """

    def __init__(self, cfg: MpcTrajectoryConfig) -> None:
        if not cfg.mpc_yaml_path:
            raise ValueError('MpcTrajectoryController: mpc_yaml_path must be provided.')
        self._cfg = cfg
        self._mpc: Optional[MPC] = None
        self._control_period = 0.01
        self._dt_horizon = 0.05
        self._horizon_steps = 0
        self._last_solve_us = 0.0
        self._name = 'MpcTrajectoryController'

    @staticmethod
    def load_config_from_yaml(path: str) -> MpcTrajectoryConfig:
        """Validate the YAML file is readable and store its path.

        The full configuration is applied later by
        :func:`configure_mpc_from_yaml` during :meth:`initialize`.
        """
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            yaml.safe_load(f)
        return MpcTrajectoryConfig(mpc_yaml_path=path)

    def initialize(self, initial_state: State, example_cfg: ExampleConfig) -> None:
        if example_cfg.mpc_dt <= 0.0:
            raise ValueError('MpcTrajectoryController: example_cfg.mpc_dt must be > 0.')

        ocp_json_file = _read_ocp_json_file(self._cfg.mpc_yaml_path)
        self._mpc = MPC(ocp_json_file)
        configure_mpc_from_yaml(self._mpc, self._cfg.mpc_yaml_path)

        self._control_period = example_cfg.mpc_dt
        self._horizon_steps = int(self._mpc.get_prediction_steps())
        self._dt_horizon = float(self._mpc.get_prediction_time_step())

    def reference_horizon_size(self) -> int:
        return self._horizon_steps + 1

    def reference_horizon_dt(self) -> float:
        return self._dt_horizon

    def control_period(self) -> float:
        return self._control_period

    def compute_command(
        self,
        state: State,
        references: List[ReferenceSample],
    ) -> ControlCommand:
        if self._mpc is None:
            raise RuntimeError(
                'MpcTrajectoryController: initialize() must be called before use.')
        expected = self._horizon_steps + 1
        if len(references) != expected:
            raise ValueError('MpcTrajectoryController: references size mismatch.')

        position = np.asarray(state.position, dtype=float)
        orientation = np.asarray(state.orientation, dtype=float)
        velocity = np.asarray(state.linear_velocity, dtype=float)

        mpc_data = self._mpc.get_data()
        mpc_data.state.position = position
        mpc_data.state.orientation = orientation
        mpc_data.state.linear_velocity = velocity

        desired_quat = _yaw_to_quat(float(references[0].yaw))
        params = mpc_data.parameters

        for k in range(self._horizon_steps + 1):
            r = references[k]
            params.set_desired_position(np.asarray(r.position, dtype=float), k)
            params.set_desired_velocity(np.asarray(r.velocity, dtype=float), k)
            params.set_desired_acceleration(np.asarray(r.acceleration, dtype=float), k)
            params.set_desired_orientation(desired_quat, k)

        t0 = time.perf_counter()
        status = int(self._mpc.solve())
        t1 = time.perf_counter()
        self._last_solve_us = (t1 - t0) * 1e6

        if status != 0:
            raise RuntimeError(
                f'MpcTrajectoryController: solver returned status {status}')

        thrust = float(mpc_data.actuation.thrust)
        rates = np.asarray(mpc_data.actuation.angular_velocity, dtype=float)
        return ControlCommand(thrust_n=thrust, angular_rate=rates)

    def required_reference_fields(self) -> ReferenceField:
        return make_mask([
            ReferenceField.POSITION,
            ReferenceField.VELOCITY,
            ReferenceField.ACCELERATION,
        ])

    def name(self) -> str:
        return self._name

    def last_solve_time_micros(self) -> float:
        return self._last_solve_us

File: examples/adapters/controllers/mpc_trajectory/src/mpc_trajectory_controller.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file mpc_trajectory_controller.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "adapters/controllers/mpc_trajectory_controller.hpp"

#include <array>
#include <chrono>
#include <cmath>
#include <stdexcept>
#include <utility>

#include "acados_mpc/acados_mpc_yaml.hpp"
#include "utils/example_config_utils.hpp"

namespace mpc_examples::adapters {

namespace {

std::array<double, 4> yawToQuatArray(double yaw) {
  const double half = 0.5 * yaw;
  return {std::cos(half), 0.0, 0.0, std::sin(half)};
}

}  // namespace

MpcTrajectoryController::MpcTrajectoryController(const Config& cfg) : cfg_(cfg) {
  if (cfg_.mpc_yaml_path.empty()) {
    throw std::invalid_argument("MpcTrajectoryController: mpc_yaml_path must be provided.");
  }
}

MpcTrajectoryController::Config MpcTrajectoryController::loadConfigFromYaml(
    const std::string& path) {
  // Validates the YAML file is readable; the full configuration is applied
  // later by acados_mpc::configureMpcFromYaml() during initialize().
  (void)detail::loadYamlRoot(path);
  Config cfg;
  cfg.mpc_yaml_path = path;
  return cfg;
}

void MpcTrajectoryController::initialize(const mav_model::State& /*initial_state*/,
                                         const ExampleConfig& example_cfg) {
  if (example_cfg.mpc_dt <= 0.0) {
    throw std::invalid_argument("MpcTrajectoryController: example_cfg.mpc_dt must be > 0.");
  }

  mpc_ = std::make_unique<acados_mpc::MPC>();
  acados_mpc::configureMpcFromYaml(*mpc_, cfg_.mpc_yaml_path);

  control_period_ = example_cfg.mpc_dt;
  horizon_steps_  = mpc_->getPredictionSteps();
  dt_horizon_     = mpc_->getPredictionTimeStep();
}

framework::ControlCommand MpcTrajectoryController::computeCommand(
    const mav_model::State& state,
    const std::vector<framework::ReferenceSample>& references) {
  if (!mpc_) {
    throw std::runtime_error("MpcTrajectoryController: initialize() must be called before use.");
  }
  const int expected = horizon_steps_ + 1;
  if (static_cast<int>(references.size()) != expected) {
    throw std::invalid_argument("MpcTrajectoryController: references size mismatch.");
  }

  const Eigen::Vector3d position       = state.getPositionVector();
  const Eigen::Quaterniond orientation = state.getOrientationVector();
  const Eigen::Vector3d velocity       = state.getLinearVelocityVector();

  acados_mpc::MPCData* mpc_data = mpc_->getData();
  mpc_data->state.setPosition({position.x(), position.y(), position.z()});
  mpc_data->state.setOrientation(
      {orientation.w(), orientation.x(), orientation.y(), orientation.z()});
  mpc_data->state.setLinearVelocity({velocity.x(), velocity.y(), velocity.z()});

  const std::array<double, 4> desired_quat = yawToQuatArray(references.front().yaw);

  for (int k = 0; k <= horizon_steps_; ++k) {
    const framework::ReferenceSample& r = references[static_cast<std::size_t>(k)];
    mpc_data->p_params.setDesiredPosition({r.position.x(), r.position.y(), r.position.z()}, k);
    mpc_data->p_params.setDesiredVelocity({r.velocity.x(), r.velocity.y(), r.velocity.z()}, k);
    mpc_data->p_params.setDesiredAcceleration(
        {r.acceleration.x(), r.acceleration.y(), r.acceleration.z()}, k);
    mpc_data->p_params.setDesiredOrientation(desired_quat, k);
  }

  const auto t0    = std::chrono::high_resolution_clock::now();
  const int status = mpc_->solve();
  const auto t1    = std::chrono::high_resolution_clock::now();
  last_solve_us_   = std::chrono::duration<double>(t1 - t0).count() * 1e6;

  if (status != 0) {
    throw std::runtime_error("MpcTrajectoryController: solver returned status " +
                             std::to_string(status));
  }

  framework::ControlCommand cmd;
  cmd.thrust_n = mpc_data->actuation.getThrust();
  const auto w = mpc_data->actuation.getAngularVelocity();
  cmd.angular_rate = {w[0], w[1], w[2]};
  return cmd;
}

}  // namespace mpc_examples::adapters
